larpandoracontent/LArMonitoring/VisualParticleMonitoringAlgorithm.cc

```diff
@@ -0,0 +1,428 @@
+ */
+
+#include "Pandora/AlgorithmHeaders.h"
+
+#include "larpandoracontent/LArMonitoring/VisualParticleMonitoringAlgorithm.h"
+
+#include "larpandoracontent/LArHelpers/LArPfoHelper.h"
+#include "larpandoracontent/LArObjects/LArCaloHit.h"
+#include "larpandoracontent/LArObjects/LArMCParticle.h"
+
+using namespace pandora;
+
+namespace lar_content
+{
+
+VisualParticleMonitoringAlgorithm::VisualParticleMonitoringAlgorithm() :
+    m_visualizeMC(false),
+    m_visualizePfo(false),
+    m_groupMCByPdg(false),
+    m_showPfoByPid(false),
+    m_showPfoMatchedMC(false),
+    m_isTestBeam{false}
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+VisualParticleMonitoringAlgorithm::~VisualParticleMonitoringAlgorithm()
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+StatusCode VisualParticleMonitoringAlgorithm::Run()
+{
+#ifdef MONITORING
+    LArMCParticleHelper::MCContributionMap targetMCParticleToHitsMap;
+    if (m_visualizeMC || m_showPfoMatchedMC)
+    {
+        const CaloHitList *pCaloHitList(nullptr);
+        PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::GetList(*this, m_caloHitListName, pCaloHitList));
+        const MCParticleList *pMCParticleList(nullptr);
+        PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::GetCurrentList(*this, pMCParticleList));
+        this->MakeSelection(pMCParticleList, pCaloHitList, targetMCParticleToHitsMap);
+    }
+
+    if (m_visualizeMC)
+    {
+        if (m_groupMCByPdg)
+            this->VisualizeMCByPdgCode(targetMCParticleToHitsMap);
+        else
+            this->VisualizeIndependentMC(targetMCParticleToHitsMap);
+    }
+    if (m_visualizePfo)
+    {
+        const PfoList *pPfoList(nullptr);
+        PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::GetList(*this, m_pfoListName, pPfoList));
+        if (m_showPfoByPid)
+        {
+            this->VisualizePfoByParticleId(*pPfoList);
+        }
+        else
+        {
+            if (m_showPfoMatchedMC)
+                this->VisualizeIndependentPfo(*pPfoList, targetMCParticleToHitsMap);
+            else
+                this->VisualizeIndependentPfo(*pPfoList);
+        }
+    }
+#endif // MONITORING
+    return STATUS_CODE_SUCCESS;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+#ifdef MONITORING
+
+void VisualParticleMonitoringAlgorithm::VisualizeIndependentMC(const LArMCParticleHelper::MCContributionMap &mcMap)
+{
+    std::map<int, const std::string> keys = {{13, "mu"}, {11, "e"}, {22, "gamma"}, {321, "kaon"}, {211, "pi"}, {2212, "p"}};
+    std::map<int, Color> colors = {{0, RED}, {1, BLACK}, {2, BLUE}, {3, CYAN}, {4, MAGENTA}, {5, GREEN}, {6, ORANGE}, {7, GRAY}};
```

Comment:
Maps could be const, although means using `.at()` below, rather than `operator[]`.